lib/perl/Genome/Ptero/Wrapper.pm

```diff
@@ -1,12 +1,198 @@
+}
+
+sub _stderr_log_path {
+    my $self = shift;
+
+    my $base_name = $self->execution->name;
+    my $output_log = File::Spec->join(abs_path($self->log_directory), "$base_name.err");
+}
+
+sub _setup_logging {
+    my $self = shift;
+
+    $self->execution->update_data(
+        stdout_log => $self->_stdout_log_path,
+        stderr_log => $self->_stderr_log_path,
+    );
+
+    open(SAVED_STDOUT, ">&STDOUT") || die "Can't save STDOUT\n";
+    open(SAVED_STDERR, ">&STDERR") || die "Can't save STDERR\n";
+
+    # redirect stdout/stderr through the annotate-log filter
+    open OUTPUT, '|-',  'annotate-log cat > ' .
+        $self->_stdout_log_path or die $!;
+    open ERROR, '|-',  'annotate-log cat > ' .
+        $self->_stderr_log_path or die $!;
+
+    STDOUT->fdopen(\*OUTPUT, 'w') or die $!;
+    STDERR->fdopen(\*ERROR,  'w') or die $!;
+}
+
+sub _teardown_logging {
+    my $self = shift;
+
+    open(STDOUT, ">&SAVED_STDOUT") || die "Can't restore STDOUT\n";
+    open(STDERR, ">&SAVED_STDERR") || die "Can't restore STDERR\n";
+}
+
+sub _log_execution_information {
+    my $self = shift;
+
+    $self->status_message("COMMAND: PTERO_WORKFLOW_EXECUTION_URL=%s %s " .
+        "ptero wrapper --command-class=\"%s\" --method=\"%s\" --log-directory=\"%s\"",
+        $ENV{PTERO_WORKFLOW_EXECUTION_URL}, $0,
+        $self->command_class, $self->method, abs_path($self->log_directory));
+}
+
+sub _instantiate_command {
+    my ($self, $inputs) = @_;
+
+    my $pkg = $self->command_class;
+    eval "use $pkg";
+    my $cmd = eval {$pkg->create(%$inputs)};
+    my $error = $@;
+    if ($error) {
+        Carp::confess sprintf(
+            "Failed to intantiate class (%s) with inputs (%s): %s",
+            $pkg, Data::Dump::pp($inputs), $error);
+    }
+    return $cmd;
+}
+
+sub _run_command {
+    my ($self, $command) = @_;
+
+    my $method = $self->method;
+    my $ret = eval { $command->$method() };
+    my $error = $@;
+    if ($error) {
+        Carp::confess sprintf(
+            "Crashed in %s for command %s: %s",
+            $self->method, $self->command_class, $error,
+        );
+    }
+    unless ($ret) {
+        Carp::confess sprintf("Failed to %s for command %s.",
+            $self->method, $self->command_class,
+        );
+    }
+
+    _commit();
+
+    $self->status_message("Succeeded to %s command %s",
+        $method, $self->command_class);
+}
+
+sub _commit {
+    eval {
+        UR::Context->commit();
+    };
+
+    my $error = $@;
+    if ($error) {
+        Carp::confess "Failed to commit: $error";
+    }
+}
+
+sub _get_command_outputs {
+    my ($cmd, $pkg) = @_;
+
+    my %outputs;
+    for my $prop (_output_names($pkg)) {
+        my $prop_name = $prop->property_name;
+        my $value = $prop->is_many ? [$cmd->$prop_name] : $cmd->$prop_name;
+        $outputs{$prop_name} = $value;
+    }
+
+    $outputs{result} = 1 unless exists $outputs{result};
```

Comment:
This was brought over from flow which supported event based workflows, I think it can go now.